infrastructure/logger/backend.go

```diff
@@ -0,0 +1,187 @@
+type Backend struct {
+	flag      uint32
+	isRunning uint32
+	writers   []logWriter
+	writeChan chan logEntry
+	syncClose sync.Mutex // used to sync that the logger finished writing everything
+}
+
+// NewBackendWithFlags configures a Backend to use the specified flags rather than using
+// the package's defaults as determined through the LOGFLAGS environment
+// variable.
+func NewBackendWithFlags(flags uint32) *Backend {
+	return &Backend{flag: flags, writeChan: make(chan logEntry, logsBuffer)}
+}
+
+// NewBackend creates a new logger backend.
+func NewBackend() *Backend {
+	return NewBackendWithFlags(defaultFlags)
+}
+
+const (
+	defaultThresholdKB = 100 * 1000 // 100 MB logs by default.
+	defaultMaxRolls    = 8          // keep 8 last logs by default.
+)
+
+type logWriter interface {
+	io.WriteCloser
+	LogLevel() Level
+}
+
+type logWriterWrap struct {
+	io.WriteCloser
+	logLevel Level
+}
+
+func (lw logWriterWrap) LogLevel() Level {
+	return lw.logLevel
+}
+
+// AddLogFile adds a file which the log will write into on a certain
+// log level with the default log rotation settings. It'll create the file if it doesn't exist.
+func (b *Backend) AddLogFile(logFile string, logLevel Level) error {
+	return b.AddLogFileWithCustomRotator(logFile, logLevel, defaultThresholdKB, defaultMaxRolls)
+}
+
+// AddLogWriter adds a type implementing io.WriteCloser which the log will write into on a certain
+// log level with the default log rotation settings. It'll create the file if it doesn't exist.
+func (b *Backend) AddLogWriter(logWriter io.WriteCloser, logLevel Level) error {
+	if b.IsRunning() {
+		return errors.New("The logger is already running")
+	}
+	b.writers = append(b.writers, logWriterWrap{
+		WriteCloser: logWriter,
+		logLevel:    logLevel,
+	})
+	return nil
+}
+
+// AddLogFileWithCustomRotator adds a file which the log will write into on a certain
+// log level, with the specified log rotation settings.
+// It'll create the file if it doesn't exist.
+func (b *Backend) AddLogFileWithCustomRotator(logFile string, logLevel Level, thresholdKB int64, maxRolls int) error {
+	if b.IsRunning() {
+		return errors.New("The logger is already running")
+	}
+	logDir, _ := filepath.Split(logFile)
+	// if the logDir is empty then `logFile` is in the cwd and there's no need to create any directory.
+	if logDir != "" {
+		err := os.MkdirAll(logDir, 0700)
+		if err != nil {
+			return errors.Errorf("failed to create log directory: %+v", err)
+		}
+	}
+	r, err := rotator.New(logFile, thresholdKB, false, maxRolls)
+	if err != nil {
+		return errors.Errorf("failed to create file rotator: %s", err)
+	}
+	b.writers = append(b.writers, logWriterWrap{
+		WriteCloser: r,
+		logLevel:    logLevel,
+	})
+	return nil
+}
+
+// Run launches the logger backend in a separate go-routine. should only be called once.
+func (b *Backend) Run() error {
+	if !atomic.CompareAndSwapUint32(&b.isRunning, 0, 1) {
+		return errors.New("The logger is already running")
+	}
+	go func() {
```

Comment:
This is the main loop of the logger, I'd give it the honor of it's own function.
(Or maybe not this function, but the `for log := range b.writeChan` loop itself, to keep all the boilerplate out of it)